File: chapter1/5.py
'''
There are three types of edits that can be performed on strings: insert a character,
remove a character, or replace a character. Given two strings, write a function to
check if they are one edit (or zero edits) away.

'''
def is_one_edit_away(a,b):
	size_difference = abs(len(a)-len(b))
	if(size_difference>1): return False

	i = 0
	j = 0
	found_discrepancy = False
	while(i<len(a) and j<len(b)):
		if a[i]!=b[j]:
			if size_difference == 0:
				if found_discrepancy:
					return False
				else:
					found_discrepancy = True
			else:
				if found_discrepancy:
					return False
				else:
					if(len(a)>len(b)):
						i += 1
					else:
						j += 1
						found_discrepancy = True
		i += 1
		j += 1

	return True

def main():
	print(is_one_edit_away("aba", "aaa"))

if __name__ == "__main__":
	main()

File: chapter1/c++/5.cpp
/*
There are three types of edits that can be performed on strings: insert a character, remove a character, or replace a character. Given two strings, write a function to check if they are one edit (or zero edits) away.
*/

#include <bits/stdc++.h>
using namespace std;

bool isOneEditAway(string a, string b) {
	int sizeDifference = abs((int)a.size()-(int)b.size());
	if(sizeDifference>1) return false;

	int i = 0;
	int j = 0;
	bool foundDiscrepancy = false;
	while(i<a.size() && j<b.size()) {
		if(a[i]!=b[j]) {
			if(sizeDifference==0) {
				if(foundDiscrepancy) {
					return false;
				} else {
					foundDiscrepancy = true;
				}
			} else {
				if(foundDiscrepancy) {
					return false;
				} else {
					if(a.size()>b.size()) i++;
					else j++;
					foundDiscrepancy = true;
				}
			}
		}
		i++;
		j++;
	}
	return true;
}


int main() {
	cout << (isOneEditAway("aabrdjaad", "aabrdkjaad") ? "TRUE" : "FALSE") << endl;
}
